Refactor RPC matchBuiltInOp to get rid of exception swallowing

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49009 Refactor RPC matchBuiltInOp to get rid of exception swallowing**

As per the title, we should generally not have exception swalling and
this commit makes it so that if there is a true error in JIT operator
resolution, it is propagated back to the RPC callee and we don't silently
swallow any other exceptions that may happen. Swallowing the exceptions
previously resulted in hard to debug issues such as unexpected ops showing up
in profiler, and flaky tests which were fixed by
https://github.com/pytorch/pytorch/pull/41287

Added a unittest that validates the error that comes from `jit/pybind_utils.h`.

Differential Revision: [D25392905](https://our.internmc.facebook.com/intern/diff/D25392905/)

rohan-varma committed:
Refactor RPC matchBuiltInOp to get rid of exception swallowing

As per the title, we should generally not have exception swalling and
this commit makes it so that if there is a true error in JIT operator
resolution, it is propagated back to the RPC callee and we don't silently
swallow any other exceptions that may happen. Swallowing the exceptions
previously resulted in hard to debug issues such as unexpected ops showing up
in profiler, and flaky tests which were fixed by
https://github.com/pytorch/pytorch/pull/41287

Added a unittest that validates the error that comes from `jit/pybind_utils.h`.

Differential Revision: [D25392905](https://our.internmc.facebook.com/intern/diff/D25392905/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Refactor RPC matchBuiltInOp to get rid of exception swallowing"


As per the title, we should generally not have exception swalling and
this commit makes it so that if there is a true error in JIT operator
resolution, it is propagated back to the RPC callee and we don't silently
swallow any other exceptions that may happen. Swallowing the exceptions
previously resulted in hard to debug issues such as unexpected ops showing up
in profiler, and flaky tests which were fixed by
https://github.com/pytorch/pytorch/pull/41287

Added a unittest that validates the error that comes from `jit/pybind_utils.h`.

Differential Revision: [D25392905](https://our.internmc.facebook.com/intern/diff/D25392905/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Refactor RPC matchBuiltInOp to get rid of exception swallowing"


As per the title, we should generally not have exception swalling and
this commit makes it so that if there is a true error in JIT operator
resolution, it is propagated back to the RPC callee and we don't silently
swallow any other exceptions that may happen. Swallowing the exceptions
previously resulted in hard to debug issues such as unexpected ops showing up
in profiler, and flaky tests which were fixed by
https://github.com/pytorch/pytorch/pull/41287

Added a unittest that validates the error that comes from `jit/pybind_utils.h`.

Differential Revision: [D25392905](https://our.internmc.facebook.com/intern/diff/D25392905/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Refactor RPC matchBuiltInOp to get rid of exception swallowing"


As per the title, we should generally not have exception swalling and
this commit makes it so that if there is a true error in JIT operator
resolution, it is propagated back to the RPC callee and we don't silently
swallow any other exceptions that may happen. Swallowing the exceptions
previously resulted in hard to debug issues such as unexpected ops showing up
in profiler, and flaky tests which were fixed by
https://github.com/pytorch/pytorch/pull/41287

Added a unittest that validates the error that comes from `jit/pybind_utils.h`.

Differential Revision: [D25392905](https://our.internmc.facebook.com/intern/diff/D25392905/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Refactor RPC matchBuiltInOp to get rid of exception swallowing"


As per the title, we should generally not have exception swalling and
this commit makes it so that if there is a true error in JIT operator
resolution, it is propagated back to the RPC callee and we don't silently
swallow any other exceptions that may happen. Swallowing the exceptions
previously resulted in hard to debug issues such as unexpected ops showing up
in profiler, and flaky tests which were fixed by
https://github.com/pytorch/pytorch/pull/41287

Added a unittest that validates the error that comes from `jit/pybind_utils.h`.

Differential Revision: [D25392905](https://our.internmc.facebook.com/intern/diff/D25392905/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -61,36 +61,42 @@ std::shared_ptr<Operator> matchBuiltinOp(
+    for (const auto& op : ops) {
       if (op->isC10Op()) {
-        return op;
+        c10OpsForSymbol.push_back(op);
```
is it possible to use a vector iterator to remove the op from `ops` so that we don't need to go through c10 ops twice?

mrshenli commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -869,6 +869,15 @@ inline IValue argumentToIValue(
             py::repr(object)),
         "\nCast error details: ",
         error.what()));
+  } catch (const py::error_already_set& error) {
```
when will we hit this new path?

rohan-varma replied:
In operator matching from RPC, we can get errors such as `TypeError: 'int' object is not iterable` when we try to match operators that are not a match. We were previously swallowing these exceptions and handling them with VLOG in the RPC code, but they are errors during schema match, so we convert them appropriately to schema match errors.

Ideally, this would surface as type error but `catch (const py::type_error)` does not catch the error, but py::error_already_set does catch it.

mrshenli reviewed (approved):
LGTM! Thanks for fixing!!

wanchaol commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -1235,20 +1244,19 @@ inline py::object invokeScriptMethodFromPython(
 }
 
-inline py::object invokeOperatorFromPython(
+inline std::pair<std::shared_ptr<Operator>, Stack> getOpWithStack(
```
I am ok with changing the return type to bundle with Stack, but the fact that we can introducing a flag purely for ignoring a error message make me feels we are doing it in a wrong way.  maybe we should make the c10 op special handling (early return) inside this function directly?

rohan-varma replied:
I agree with the sentiment that introducing a flag here to control whether we throw or not is not ideal. Although, I am a bit reluctant to add the c10 op special-casing here, in more general JIT code that is used elsewhere besides RPC.

The order of operators seems to be somewhat brittle and differs in different build modes (see internal discussion), which is why that special casing exists in the first place. Given that, until we can remove it completely I'd like to keep it only in RPC codebase to avoid this logic changing the JIT too much. Maybe we can remove the flag from here, and then special case the RPC code to retry (kind of similar to what we do currently in this diff).

wanchaol commented on torch/csrc/distributed/rpc/python_functions.cpp:
```diff
@@ -61,36 +61,42 @@ std::shared_ptr<Operator> matchBuiltinOp(
+            c10OpsForSymbol, args, kwargs, false /* throwOnMatchFailure */);
+
+    if ((matchedOperator = std::get<0>(opWithStack)) != nullptr) {
+      stack = std::get<1>(opWithStack);
```
the stack seems not getting used at all?

rohan-varma replied:
We take in the stack by reference into this function and mutate it, in the same way that the function did before this diff.

wanchaol reviewed (commented):
Please see inline comments. Mostly looks good to me, but would like to get more clear signal on how Stack is getting used, and if we could have a better way to avoid calling this function twice (just to ignore error message for some ops).

rohan-varma commented on torch/csrc/jit/python/pybind_utils.h:
```diff
@@ -1270,10 +1278,17 @@ inline py::object invokeOperatorFromPython(
       throw std::runtime_error(ss.str());
     }
 
-    pybind11::gil_scoped_release no_gil_guard;
```
Good catch, looks like I missed this. We should definitely release GIL as we just execute the found operation in C++, there is no need for GIL.

rohan-varma committed:
Update on "Refactor RPC matchBuiltInOp to get rid of exception swallowing"


As per the title, we should generally not have exception swalling and
this commit makes it so that if there is a true error in JIT operator
resolution, it is propagated back to the RPC callee and we don't silently
swallow any other exceptions that may happen. Swallowing the exceptions
previously resulted in hard to debug issues such as unexpected ops showing up
in profiler, and flaky tests which were fixed by
https://github.com/pytorch/pytorch/pull/41287

Added a unittest that validates the error that comes from `jit/pybind_utils.h`.

Differential Revision: [D25392905](https://our.internmc.facebook.com/intern/diff/D25392905/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Refactor RPC matchBuiltInOp to get rid of exception swallowing"


As per the title, we should generally not have exception swalling and
this commit makes it so that if there is a true error in JIT operator
resolution, it is propagated back to the RPC callee and we don't silently
swallow any other exceptions that may happen. Swallowing the exceptions
previously resulted in hard to debug issues such as unexpected ops showing up
in profiler, and flaky tests which were fixed by
https://github.com/pytorch/pytorch/pull/41287

Added a unittest that validates the error that comes from `jit/pybind_utils.h`.

Differential Revision: [D25392905](https://our.internmc.facebook.com/intern/diff/D25392905/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Refactor RPC matchBuiltInOp to get rid of exception swallowing"


As per the title, we should generally not have exception swalling and
this commit makes it so that if there is a true error in JIT operator
resolution, it is propagated back to the RPC callee and we don't silently
swallow any other exceptions that may happen. Swallowing the exceptions
previously resulted in hard to debug issues such as unexpected ops showing up
in profiler, and flaky tests which were fixed by
https://github.com/pytorch/pytorch/pull/41287

Added a unittest that validates the error that comes from `jit/pybind_utils.h`.

Differential Revision: [D25392905](https://our.internmc.facebook.com/intern/diff/D25392905/)

[ghstack-poisoned]

wanchaol reviewed (approved):
looks good to me, thanks!

rohan-varma committed:
Update on "Refactor RPC matchBuiltInOp to get rid of exception swallowing"


As per the title, we should generally not have exception swalling and
this commit makes it so that if there is a true error in JIT operator
resolution, it is propagated back to the RPC callee and we don't silently
swallow any other exceptions that may happen. Swallowing the exceptions
previously resulted in hard to debug issues such as unexpected ops showing up
in profiler, and flaky tests which were fixed by
https://github.com/pytorch/pytorch/pull/41287

Added a unittest that validates the error that comes from `jit/pybind_utils.h`.

Differential Revision: [D25392905](https://our.internmc.facebook.com/intern/diff/D25392905/)

[ghstack-poisoned]

rohan-varma committed:
Update on "Refactor RPC matchBuiltInOp to get rid of exception swallowing"


As per the title, we should generally not have exception swalling and
this commit makes it so that if there is a true error in JIT operator
resolution, it is propagated back to the RPC callee and we don't silently
swallow any other exceptions that may happen. Swallowing the exceptions
previously resulted in hard to debug issues such as unexpected ops showing up
in profiler, and flaky tests which were fixed by
https://github.com/pytorch/pytorch/pull/41287

Added a unittest that validates the error that comes from `jit/pybind_utils.h`.

Differential Revision: [D25392905](https://our.internmc.facebook.com/intern/diff/D25392905/)

[ghstack-poisoned]
